File: mobile/src/main/java/treehou/se/habit/module/RxPresenter.java
package treehou.se.habit.module;


import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import com.trello.rxlifecycle.LifecycleTransformer;
import com.trello.rxlifecycle.RxLifecycle;

import rx.Observable;
import rx.subjects.BehaviorSubject;
import treehou.se.habit.module.RxPresenterEvent.PresenterEvent;
import treehou.se.habit.mvp.BasePresenter;

public class RxPresenter implements BasePresenter {

    private final BehaviorSubject<PresenterEvent> lifecycleSubject = BehaviorSubject.create();

    @NonNull
    @CheckResult
    public final Observable<PresenterEvent> lifecycle() {
        return lifecycleSubject.asObservable();
    }

    @NonNull
    @CheckResult
    public final <K> LifecycleTransformer<K> bindUntilEvent(@NonNull PresenterEvent event) {
        return RxLifecycle.bindUntilEvent(lifecycleSubject, event);
    }

    @NonNull
    @CheckResult
    public final <K> LifecycleTransformer<K> bindToLifecycle() {
        return RxLifecyclePresenter.bindPresenter(lifecycleSubject);
    }

    @Override
    @CallSuper
    public void load(Bundle savedData) {
        lifecycleSubject.onNext(PresenterEvent.LOAD);
    }

    @Override
    @CallSuper
    public void subscribe() {
        lifecycleSubject.onNext(PresenterEvent.SUBSCRIBE);
    }

    @Override
    @CallSuper
    public void unsubscribe() {
        lifecycleSubject.onNext(PresenterEvent.UNSUBSCRIBE);
    }

    @Override
    @CallSuper
    public void unload() {
        lifecycleSubject.onNext(PresenterEvent.UNLOAD);
    }

    @Override
    public void save(Bundle savedData) {
    }
}

File: mobile/src/main/java/treehou/se/habit/dagger/RxPresenter.kt
package treehou.se.habit.dagger


import android.os.Bundle
import android.support.annotation.CallSuper
import android.support.annotation.CheckResult

import com.trello.rxlifecycle2.LifecycleTransformer
import com.trello.rxlifecycle2.RxLifecycle

import io.reactivex.BackpressureStrategy
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import treehou.se.habit.dagger.RxPresenterEvent.PresenterEvent
import treehou.se.habit.mvp.BasePresenter

open class RxPresenter : BasePresenter {

    private val lifecycleSubject = BehaviorSubject.create<PresenterEvent>()

    @CheckResult
    fun lifecycle(): Observable<PresenterEvent> {
        return lifecycleSubject.toFlowable(BackpressureStrategy.BUFFER).toObservable()
    }

    @CheckResult
    fun <K> bindUntilEvent(event: PresenterEvent): LifecycleTransformer<K> {
        return RxLifecycle.bindUntilEvent(lifecycleSubject, event)
    }

    @CheckResult
    fun <K> bindToLifecycle(): LifecycleTransformer<K> {
        return RxLifecyclePresenter.bindPresenter(lifecycleSubject)
    }

    @CallSuper
    override fun load(launchData: Bundle?, savedData: Bundle?) {
        lifecycleSubject.onNext(PresenterEvent.LOAD)
    }

    @CallSuper
    override fun subscribe() {
        lifecycleSubject.onNext(PresenterEvent.SUBSCRIBE)
    }

    @CallSuper
    override fun unsubscribe() {
        lifecycleSubject.onNext(PresenterEvent.UNSUBSCRIBE)
    }

    @CallSuper
    override fun unload() {
        lifecycleSubject.onNext(PresenterEvent.UNLOAD)
    }

    override fun save(savedData: Bundle?) {}
}
